dashboard/app/access.go

```diff
@@ -59,6 +59,24 @@ func emailInAuthDomains(email string, authDomains []string) bool {
 	return false
 }
 
+func currentUser(c context.Context, r *http.Request) *user.User {
+	u := user.Current(c)
+	if u != nil {
+		return u
+	}
+	// In case of the wrong token let's silently give public access.
+	// Expired tokens will enable throttling.
+	u, _ = user.CurrentOAuth(c, "https://www.googleapis.com/auth/userinfo.email")
```

Comment:
accessLevel already gives AccessPublic access if u==nil. What I am missing here?

Reply:
I'd like the "u" to not be nil if we can authenticate user by "Authorization: Bearer ..."  token.

Reply:
user.Current(c) authenticates users differently. Internally I see it uses many AppEngine specific headers.

Reply:
You're probably confused by the comment in line 67. Let me rephrase it.

Reply:
It is used in #5378 .

Reply:
New auth method makes me nervous.
Are you sure this is doing the right thing? (for gmail.com AuthDomain user logged in into the account)
Are positive/negative tests for this possible?

Reply:
I meant automated tests like we have in access_test.go.
If somebody later refactors this code, the tests need to be repeated.

Reply:
We can test for the feature presence.
I thought you're worrying about the potential security breach here. I don't think we can automate this part.

Reply:
> I thought you're worrying about the potential security breach here. 

Yes, that's what I am worried about.

Reply:
> I don't think we can automate this part.

What options have you considered?
Here Greg #2638, #2689, #2694 added some tests for OAuth.

Reply:
Thanks for the links. I forgot about the fact dev_appserver may support oauth2 tokens.
I'm not sure it supports them but it makes sense to double check..

1. All the header parsing and verification logic is inside of  the user.CurrentOAuth. To check it, AppEngine deployment was my the only good option. But we don't have that level integration testing yet.
2. Alternatively we can add TestAccessLevel() to verify "accessLevel()" but it will work as a smoke test to check the feature presence. It'll not check the feature integration.
3. One more alternative is to mock the user.CurrentOAuth but what should be inside? Parsing + verification, some 3rd party oauth verification, fake server with custom key chain? All these look too complex.

Reply:
I don't have time to deep dive into this right now.